Implement Symbolic RVs

ricardoV94 opened the pull request:
This PR wraps the graphs returned by the old `SymbolicDistribution.rv_op` in a `SymbolicRV` `OpFromGraph` `Op` so that they can be manipulated like vanilla `RandomVariable`s. The main benefit is that we can now:
1. Change their size, which is necessary for creating graphs with nested `SymbolicRV`s (nested mixtures, censored mixtures, AR with censored init_dist, you name it)
2.  Dispatch moment functions, which we noted was necessary in #5762

When these `OpFromGraph` don't have a specialized logprob dispatched, they can be inlined inside Aeppl, so that their logprob can be derived dynamically. This was one of the main reasons for creating `SymbolicDistribution`s in the first place.

Closes #5533 

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## Bugfixes / New features
- Implemented general univariate RandomWalk distribution factory
- Distribution factories like Mixture, Censored, RandomWalk, and AR can now be composed together
- Reintroduced support for nested Mixtures

## Docs / Maintenance
- Implemented `SymbolicRV`s to encapsulate graphs from distribution factories
- Removed temporary `SymbolicDistribution` class in favor of `Distribution`
- `change_rv_size` was removed in favor of the more general `change_dist_size`
- `_LKJCholeskyCov` and `GaussianRandomWalk` are now `SymbolicRV`s

ricardoV94 committed:
Wrap SymbolicDistribution graphs in SymbolicRandomVariable Ops

This allows us to treat these variables as if they were pure RandomVariables (dispatch moments, resize, etc.)

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

ricardoV94 committed:
Refactor change_rv_size into a dispatch-based method

* change_rv_size is now called change_dist_size

ricardoV94 committed:
Enable nested SymbolicRandomVariables

ricardoV94 committed:
Deprecate SymbolicDistribution in favor of Distribution class

* Also fixes index error in AR when init_dist is scalar and size is None
* Remove unused rv_class attribute

lucianopaz commented on pymc/distributions/censored.py:
```diff
@@ -100,26 +107,23 @@ def rv_op(cls, dist, lower=None, upper=None, size=None):
+        return CensoredRV(
+            inputs=[dist_, lower_, upper_],
+            outputs=[censored_rv_],
+            ndim_supp=0,
```
Is it necessary to restrict the ndim_supp only to scalar distributions?

ricardoV94 replied:
It's not, but for now we only have logcdf for scalar distributions, and aeppl and pymc check the `dist` is scalar so I just hard-coded it here.

lucianopaz commented on pymc/distributions/censored.py:
```diff
@@ -100,26 +107,23 @@ def rv_op(cls, dist, lower=None, upper=None, size=None):
 
 
-@_moment.register(Clip)
+@_moment.register(CensoredRV)
```
This is so much cleaner now!

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -464,7 +468,7 @@ def __new__(
         if resize_shape_from_dims:
             resize_size_from_dims = find_size(
-                shape=resize_shape_from_dims, size=None, ndim_supp=rv_out.tag.ndim_supp
+                shape=resize_shape_from_dims, size=None, ndim_supp=rv_out.owner.op.ndim_supp
```
Why was this tag before??

ricardoV94 replied:
It was just a temporary work-around  when cleaning #6063

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -542,15 +546,45 @@ def dist(
+# Registered before pre-canonicalization which happens at position=-10
+logprob_rewrites_db.register(
+    "inline_SymbolicRandomVariable",
+    in2out(inline_SymbolicRandomVariable),
```
What does in2out do?

ricardoV94 replied:
It's the simplest (maybe?) form of applying a rewrite, it just traverses a graph from inputs to outputs and tries to apply the rewrite in nodes that match the `tracks` Op

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -542,15 +546,45 @@ def dist(
+    "inline_SymbolicRandomVariable",
+    in2out(inline_SymbolicRandomVariable),
+    "basic",
+    position=-20,
```
I don’t really understand what the rewrite_db arguments mean. Could you explain these to me?

ricardoV94 replied:
I can add kwargs to make it more readable. It's just name, rewrite, tag (used for manually including or excluding rewrites), and order in which rewrite is applied from all the rewrites in the database

ricardoV94 replied:
Actually cannot use kwargs for this one.

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -542,15 +546,45 @@ def dist(
+
+    def __init__(self, *args, ndim_supp, **kwargs):
+        self.ndim_supp = ndim_supp
+        kwargs.setdefault("inline", True)
```
What happens if inline is false?

ricardoV94 replied:
When inline is False the inner code of an OpFromGraph remains as a black-box, so compilation does not mix (or optimize) across the outer and inner nodes. This only matters for the generative/random functions, because in the logp functions, the whole node is replaced by the logp expression instead.

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -861,3 +865,77 @@ def default_moment(rv, size, *rv_inputs, rv_name=None, has_fallback=False, ndim_
         )
+
+
+def change_dist_size(
```
Shouldn’t this live on shape_utils?

ricardoV94 replied:
Probably

ricardoV94 replied:
Done

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -861,3 +865,77 @@ def default_moment(rv, size, *rv_inputs, rv_name=None, has_fallback=False, ndim_
+    new_size
+        The new size.
+    expand:
+        Expand the existing size by `new_size`.
```
I can’t really understand the docstring for this. Could you add an example that shows what expands will do? I had to read the source code to understand it

ricardoV94 replied:
Sure

ricardoV94 replied:
I should also add a note that the original RV should not be used anymore, as the returned one might use the same RandomState/Generator variable as the original one.

ricardoV94 replied:
Updated the docstrings, what do you think?

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -861,3 +865,77 @@ def default_moment(rv, size, *rv_inputs, rv_name=None, has_fallback=False, ndim_
+    return new_rv
+
+
+@_change_dist_size.register(SpecifyShape)
```
Should this also dispatch asserts and prints?

ricardoV94 replied:
I never used prints so I don't know. 

I am not sure about asserts, what if they are related to shapes of inputs/outputs? We can't easily know what they are about. If a variable requires Asserts to work properly, I would suggest they are added in the inner graph of a SymbolicRV. The custom `_change_dist_size` dispatch function for that SymbolicRV will then know how to modify/recreate the asserts needed in the resized variable.

lucianopaz reviewed (commented):
Very nice work Ricardo! I skimmed through things and left a bunch of comments. Once you think this isn’ta draft anymore I’ll do a more thorough review of the tests. Did anything change in aeppl or aesara to enable this?

ricardoV94 commented:
> Once you think this isn’t a draft anymore I’ll do a more thorough review of the tests. 

I'll re-request your review then.

> Did anything change in aeppl or aesara to enable this?

Nope. It's all PyMC insfrastructure.

larryshamalama commented on pymc/distributions/distribution.py:
```diff
@@ -150,35 +155,34 @@ def fn(*args, **kwargs):
+    return rv_out, resize_shape_from_dims
 
 
 class Distribution(metaclass=DistributionMeta):
```
Unrelated to this PR: what does the class attribute `rv_class` do?

ricardoV94 replied:
Nothing?

ricardoV94 replied:
Removed it

larryshamalama commented on pymc/distributions/distribution.py:
```diff
@@ -343,223 +350,49 @@ def dist(
+    op = node.op
+    # If `Op` is not a `MeasurableVariable`, Aeppl must derive its logp from the inner graph
+    if not isinstance(op, MeasurableVariable):
+        return clone_replace(op.inner_outputs, {u: v for u, v in zip(op.inner_inputs, node.inputs)})
```
Two questions here:

1. My understanding is that, if `op` is an instance of a `MeasurableVariable`, it will (and should) have a `logp` dispatched. Will it be of any value of have an `else` statement to check that this is effectively the case in `_logprob.registry`?
2. What does the `clone_replace` here do? Does it emulate what `OpFromGraph` would do, i.e. allowing the log-probability can extract the `logp` of outputs? Can you provide an example of such a case?

ricardoV94 replied:
1. We can
2. This rewrite just does the same as the one in Aesara when OpFromGraph is inlined. This basically includes the inner graph in the original one, so that instead of seeing `CensoredRV`, Aeppl will see the inner `Clip(...)`. That way we can use Aeppl to derive it's logp, even though in PyMC we handle it inside a SymbolicRV
 
https://github.com/aesara-devs/aesara/blob/63f52536a09b13cc55481f071ccb4fc59c9740f4/aesara/compile/builders.py#L932

ricardoV94 replied:
Added a check and test

ricardoV94 replied:
Ended up removing the else branch, otherwise it fails because of the cases where we make the inputs `Unmeasurable`...

larryshamalama commented on pymc/distributions/mixture.py:
```diff
@@ -189,7 +195,7 @@ def dist(cls, w, comp_dists, **kwargs):
             #  and resize them
             if not isinstance(dist, TensorVariable) or not isinstance(
-                dist.owner.op, RandomVariable
+                dist.owner.op, (RandomVariable, SymbolicRandomVariable)
```
Thinking out loud: do we get any identifiability problems with nested mixtures? How does posterior sampling perform when a mixture component is a mixture itself? My guess is that things will work out well. Basically, given that a mixture of mixtures is a mixture itself, I'm wondering if there are any computational pitfalls with this approach, i.e. not canonicalizing a graph of nested mixtures.

Perhaps it would a `node_rewriter` of nested mixtures into a single `MarginalMixtureRV` would be an interesting avenue to pursue later...

ricardoV94 replied:
AFAIK there is nothing special about nested mixtures that is not present in flat mixtures. The same switching issues can emerge in both cases if the components are equivalent a priori.

larryshamalama commented on pymc/printing.py:
```diff
@@ -31,40 +35,62 @@
 ]
 
 
```
I'm guessing that this would close #5793 😅

ricardoV94 replied:
Yeah, that's another benefit of using SymbolicRVs

larryshamalama commented on pymc/tests/test_sampling.py:
```diff
@@ -1606,7 +1606,7 @@ def test_linear_model(self):
             mu = pm.Deterministic("mu", alpha + beta * x)
             sigma = pm.HalfNormal("sigma", 0.1)
-            obs = pm.Normal("obs", mu, sigma, observed=y)
+            obs = pm.Normal("obs", mu, sigma, observed=y, shape=x.shape)
```
Not super familiar with shape inference. Why is it necessary to provide a `shape` argument here?

ricardoV94 replied:
This is part of the previous PR #6063. You can read the discussion there.

larryshamalama reviewed (commented):
This is amazing @ricardoV94! 🙂 I feel like having an explicit `SymbolicRandomVariable` class would make the code structure of symbolic distributions much more clear. 

Would #5931 need to be first merged before this PR?

ricardoV94 commented:
> Would https://github.com/pymc-devs/pymc/pull/5931 need to be first merged before this PR?

I don't think so

ricardoV94 commented:
The failing `test_profile_variable` is odd, and looks unrelated.

ricardoV94 committed:
Avoid reusing RNGs across distinct RandomVariables

This risk is present when resizing a RandomVariable, whose new size depends on the size of the original RandomVariable. This can lead to wrong update expressions for the reused RNG variables that still depended on the original RandomVariable.

Similarly, there is a risk when clone-replacing RandomVariables, as the cloned variables will share the same RNGs. This happened when attempting to use Metropolis with Simulator variables.

* change_rv_size does not reuse old RNG
* compile_pymc raises if distinct update expression are inferred for the same RNG

lucianopaz commented on pymc/distributions/distribution.py:
```diff
@@ -355,205 +362,58 @@ def dist(
+    op = node.op
+    # If `Op` is not a `MeasurableVariable`, Aeppl must derive its logp from the inner graph
+    if not isinstance(op, MeasurableVariable):
+        return clone_replace(op.inner_outputs, {u: v for u, v in zip(op.inner_inputs, node.inputs)})
```
Could clone_replace be dangerous? It might be cloning the input nodes and end up breaking the conditional dependence between input variables and outputs of the cloned graph

ricardoV94 replied:
I am just copying what the Aesara rewrite does so I assume it's safe: 

https://github.com/aesara-devs/aesara/blob/d7fb940264a392857781b075e1b8178c603001a2/aesara/compile/builders.py#L932

ricardoV94 committed:
Convert _LKJCholeskyCov into a SymbolicRandomVariable

This is necessary because the distribution is in fact a "Factory distribution" that needs to be able to resize `sd_dist` dynamically in order to work correctly.

ricardoV94 committed:
Convert GaussianRandomWalk into a SymbolicRandomVariable

* Implements distribution agnostic univariate RandomWalk

ricardoV94 commented:
This PR has grown considerably in scope as I attempted to patch issues with `LKJCholeskyCov` and `GaussianRandomWalk`. Those are now `SymbolicRandomVariable`s as well, which makes sense as they are distribution factories (i.e., they must be composed with `init_dist`/`sd_dist`).

The nice thing to come out of this is that now we also have support for arbitrary univariate random walks (support for multivariate will be added later), as that was simpler than just converting the `GaussianRandomWalk`

Most likely some tests will still fail, and I have marked the PR has a draft, since I need to add docstrings and tests for the new `RandomWalk`

ricardoV94 committed:
Import EulerMaruyama

ricardoV94 committed:
Handle non-RandomVariables in compile_forward_sampling_function

ricardoV94 commented:
Tests are finally passing (except RTD).

I still have to add docstrings for `RandomWalk` and some tests as well, but no changes in functionality. So feel free to give another round of review :)

michaelosthege commented on pymc/aesaraf.py:
```diff
@@ -875,6 +815,53 @@ def largest_common_dtype(tensors):
+    This overrides the implementation in Aeppl, to constant fold the shapes
+    of the base vars so that RandomVariables do not show up in the logp graph,
+    which is a requirement enforced by `pymc.distributions.logprob.joint_logp`
+    """
```
This function could use some comments that explain why the steps are about.

For example, I have never seen `at.switch`, so a priori I have no clue what this does..

ricardoV94 replied:
I assume you meant `at.split`? It undoes `at.join`, just like Numpy: https://numpy.org/doc/stable/reference/generated/numpy.split.html

ricardoV94 replied:
TODO: Test this function more directly

ricardoV94 replied:
Added a direct test, hopefully it's obvious what it does with it

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -355,205 +363,66 @@ def dist(
-            functools.partial(str_for_symbolic_dist, formatting="latex"), rv_out
-        )
+    ndim_supp: int = None
+    inline_aeppl: bool = False
```
```suggestion
    inline_aeppl: bool = False
    """Specifies whether the logprob function is derived automatically by introspection of the inner graph.
    
    If `False`, a logprob function must be dispatched directly to the subclass.
    """
```

It would help to have such docstrings on the other attributes too..

ricardoV94 replied:
I already included the explanation on the docstrings above. Is there a standard for explaining attributes, like the `parameters` in numpy docstrings?

michaelosthege replied:
I'm not sure if I understand the question correctly.

You can put docstrings on attributes by placing a `""" ... """` below them. That's the standard, it's just not well known..

ricardoV94 replied:
Didn't know. Done

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -355,205 +363,66 @@ def dist(
-            A tuple of sizes for each dimension of the new RV.
-        size : int, tuple, Variable, optional
-            For creating the RV like in Aesara/NumPy.
+MeasurableVariable.register(SymbolicRandomVariable)
```
```suggestion
# Let Aeppl know that the SymbolicRandomVariable has a logprob.
MeasurableVariable.register(SymbolicRandomVariable)
```

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -541,3 +552,101 @@ def find_size(
+    dist: TensorVariable,
+    new_size: PotentialShapeType,
+    expand: bool = False,
+):
```
```suggestion
) -> TensorVariable:
```

ricardoV94 replied:
mypy starts complaining somewhere else IIRC

ricardoV94 replied:
Done.

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -541,3 +552,101 @@ def find_size(
+
+
+@_change_dist_size.register(RandomVariable)
+def change_rv_size(op, rv, new_size, expand):
```
```suggestion
def change_rv_size(op, rv, new_size, expand) -> TensorVariable:
```

michaelosthege commented on pymc/tests/test_distributions.py:
```diff
@@ -3319,9 +3305,9 @@ def test_dist_broadcasted_by_lower_upper(self):
-        assert x.eval().shape == (10, 6)
+        # sd_dist = pm.Exponential.dist(1, size=(10, 3))
+        # x = pm.LKJCholeskyCov.dist(n=3, eta=1, sd_dist=sd_dist, size=10, compute_corr=False)
+        # assert x.eval().shape == (10, 6)
```
ToDo

ricardoV94 replied:
Thanks, it was during debug :D

michaelosthege commented on pymc/tests/test_distributions_moments.py:
```diff
@@ -98,8 +98,17 @@ def test_all_distributions_have_moments():
+        dist_module.mixture.Mixture,
+        dist_module.timeseries.RandomWalk,
+        dist_module.timeseries.AR,
+    }
```
Is it then no longer possible to get the moment of a `RandomWalk`??

And `GaussianRandomWalk` is covered by `RandomWalk`?

ricardoV94 replied:
It is, but the way this check is done, it checks for `rv_op` which is a `classmethod` in "SymbolicDistributions" and not the type on which `moment` would be dispatched upon. 

Maybe we should distinguish between `rv_op` and `create_rv_op`, which for vanilla distributions is the same. That would mean we get the `logp`/`logcdf`/`moment` dispatching working automatically for now as well...

ricardoV94 replied:
Not doing the automatic dispatching, but added a new attribute

michaelosthege commented on pymc/distributions/timeseries.py:
```diff
@@ -114,109 +116,131 @@ def get_steps(
+        if not (
+            isinstance(init_dist, at.TensorVariable)
+            and init_dist.owner is not None
+            and isinstance(init_dist.owner.op, (RandomVariable, SymbolicRandomVariable))
```
This means that `init_dist` can be a `Mixture`, right?
That's cool!

michaelosthege replied:
Could `init_dist` be a list of init distributions (with `len(init_dist) == steps`)?

I mean not right now, of course, but in principle?

ricardoV94 replied:
> `This means that init_dist can be a Mixture, right?`

Yes

> Could init_dist be a list of init distributions (with len(init_dist) == steps)?

`init_dist` should not match the number of steps, but of batched dimensions. You have one per time series, to specify at which point does it start. Maybe you meant `innovation_dist`? If so, it's the same discussion as in #6090

michaelosthege replied:
Yes that was a typo, I meant `innovation_dist`

michaelosthege commented on pymc/distributions/timeseries.py:
```diff
@@ -278,57 +294,20 @@ def dist(cls, mu=0.0, sigma=1.0, *, init_dist=None, steps=None, **kwargs) -> at.
                 UserWarning,
             )
             init_dist = Normal.dist(0, 100)
```
Shouldn't this be `Normal.dist(mu, sigma)`?

ricardoV94 replied:
No. The drift `mu` does not tell us anything about the initial point. You could easily have a timeseries with the same `mu` that starts at `-100` or at `100`. https://github.com/pymc-devs/pymc/issues/5744

michaelosthege commented on pymc/distributions/timeseries.py:
```diff
@@ -114,109 +116,131 @@ def get_steps(
+
+        # Resize input distributions
+        init_dist = change_dist_size(init_dist, size)
+        innovation_dist = change_dist_size(innovation_dist, innovation_size)
```
If this line anticipated `innovation_dist` that's already a vector, one could do random walks that have different innovations in every step, right? (e.g. via `at.stack([Normal.dist(), Uniform.dist(), Beta.dist()])`

ricardoV94 replied:
This requires more discussion: https://github.com/pymc-devs/pymc/discussions/6090

ricardoV94 commented:
Officially ready for review and merge if people agree

junpenglao commented on pymc/aesaraf.py:
```diff
@@ -1004,13 +970,20 @@ def compile_pymc(
+                update_map = {rng: random_var.owner.outputs[0]}
             else:
-                rng_updates[rng] = rng.default_update
+                update_map = {rng: rng.default_update}
```
Nit:
```
            if hasattr(rng, "default_update"):
                update_map = {rng: rng.default_update}
            else:
                update_map = {rng: random_var.owner.outputs[0]}
```

junpenglao replied:
Or if you want to get fancy:
```
if update_val := getattr(rng, "default_update", None):
    update_map = {rng: update_val}
else:
    update_map = {rng: random_var.owner.outputs[0]}
```

ricardoV94 replied:
I think walrus is not in our oldest supported Python version

junpenglao replied:
🔜 then

junpenglao commented on pymc/step_methods/metropolis.py:
```diff
@@ -1046,12 +1053,14 @@ def sample_except(limit, excluded):
 
-    logp1 = pm.CallableTensor(logp0)(inarray1)
+    logp1 = CallableTensor(logp0)(inarray1)
+    # Replace any potential duplicated RNG nodes
+    (logp1,) = replace_rng_nodes((logp1,))
```
Why is this necessary now but not before?

ricardoV94 replied:
It was already necessary but we never noticed. The only logp graphs with RVs we allow are those with Simulators so it's a very experimental edge case

junpenglao reviewed (approved):
Excellent work Ricardo! I like that everything is cleaner now.

larryshamalama commented on pymc/distributions/distribution.py:
```diff
@@ -178,11 +185,55 @@ def _make_rv_and_resize_shape_from_dims(
-    rv_class = None
-    rv_op: RandomVariable = None
+    rv_op: [RandomVariable, SymbolicRandomVariable] = None
+    rv_type: MetaType = None
```
What is the need for `rv_type`? I must be missing something...

Here is my understanding: It seems that non-symbolic distributions still get an `rv_op` but are assigned an `rv_type` when calling `Distribution`, but symbolic distributions need a `rv_op` classmethod to be manually created *without* (? not sure here) `rv_type`.

ricardoV94 replied:
The rv_type is not necessary but it was useful for `test_all_distributions_have_moments`. For traditional distributions `rv_type` is added automatically by the metaclass, it's simply the `type(rv_op)`.

This seems useful to know because it's the type on which all methods are dispatched: logp/ logcdf/ moment/ change_dist_size

larryshamalama commented on pymc/distributions/logprob.py:
```diff
@@ -320,3 +324,51 @@ def ignore_logprob(rv: TensorVariable) -> TensorVariable:
     return new_node.outputs[node.outputs.index(rv)]
+
+
+@_logprob.register(MeasurableJoin)
```
Do we expect to have more of these AePPL logprob dispatches to be overwritten? Because symbolic distributions don't have a `logp` method, I was getting confused on how the log-probabilities are being retrieved

ricardoV94 replied:
We should try to avoid as much duplicated work as possible, so I imagine it won't be that frequent.

What do you mean by symbolic distributions don't have logp methods? All except Censored right now have direct logp methods.

larryshamalama replied:
Sorry, perhaps that wasn't clear or my terminology was wrong. By `logp` method, I meant that they are defined underneath the Distribution class rather than assigned via single dispatch. The second sentence was more of a comment/concern, maybe it doesn't make too much sense...

ricardoV94 replied:
That is a shortcut we did sometime ago where the MetaDistribution will do the dispatching automatically. Some context in this discussion: https://github.com/pymc-devs/pymc/issues/5308

larryshamalama commented on pymc/distributions/distribution.py:
```diff
@@ -178,11 +185,55 @@ def _make_rv_and_resize_shape_from_dims(
+    (0 for scalar, 1 for vector, ...)
+     """
+
+    inline_aeppl: bool = False
```
Are there symbolic rvs that have `inline_aeppl` to be `True` by default?

ricardoV94 replied:
Censored does

larryshamalama replied:
I see, thanks!

larryshamalama reviewed (commented):
Asking some more questions 😅

ricardoV94 committed:
Fix name error in some BaseTestDistributionRandom

* This meant that the intended checks were not being run

ricardoV94 committed:
Remove support for deprecated Ellipsis in get_steps helper

ricardoV94 committed:
Deprecate str_for_symbolic_dist in favor of str_for_dist

* Display the dists and their inputs used by SymbolicDistributions

ricardoV94 committed:
Fix missing imports in tests
